functions/index.js

```diff
@@ -0,0 +1,135 @@
+const admin = require('firebase-admin');
+admin.initializeApp();
+
+
+// Adds a message that welcomes new users into the chat
+// when a document is created on the guest list.
+exports.addWelcomeMessages = functions.firestore
+.document('rooms/{roomId}/guests/{guestId}').onCreate((snap, context) => {
+
+  const roomID = context.params.roomId;
+  const newValue = snap.data();
+  const userName = newValue.name;
+
+  // Saves the new welcome message into the database messages collection
+  admin.firestore().collection('rooms').doc(roomID).collection('messages').add({
+    name: 'Lounge Bot',
+    profilePicUrl: '/images/LoungeLogo.png',
+    text: `${userName} joined the room! Welcome!`,
+    timestamp: admin.firestore.FieldValue.serverTimestamp(),
+  });
+  console.log('Welcome message written to database.');
+});
+
+// when a user is deleted off the guest list  or signs out
+// a leave message is inputted into the chat
+exports.addLeaveMessages = functions.firestore
+.document('rooms/{roomId}/guests/{guestId}').onDelete((snap, context) => {
+
+  const roomID = context.params.roomId;
+  const deletedValue = snap.data();
+  const userName = deletedValue.name;
+
+  admin.firestore().collection('rooms').doc(roomID).collection('messages').add({
+    name: 'Lounge Bot',
+    profilePicUrl: '/images/LoungeLogo.png',
+    text: `${userName} left the room`,
+    timestamp: admin.firestore.FieldValue.serverTimestamp(),
+  });
+  console.log('Leave message written to database.');
+});
+
+/** On update of the video playback data the chat sent notifications.
+    Users are aware when the video is paused and played
+    Also when the timestamp of the video is moved outside the sync
+    window */
+exports.updatePlayBack = functions.firestore 
+.document('rooms/{roomId}/CurrentVideo/{PlaybackData}').onUpdate((change, context) => {
+
+  const roomID = context.params.roomId;
+  console.log('video change state');
+
+  // Get an object representing the document
+  const previousValue = change.before.data();
+  const changeValue = change.after.data();
+
+  // access a particular field
+  const isPlaying = changeValue.isPlaying;
+  const previousVideoState = previousValue.isPlaying;
+  const timestampNow = changeValue.timestamp;
+  const timestampBefore = previousValue.timestamp;
+  const previousSpeed = previousValue.videoSpeed;
+  const videoSpeed = changeValue.videoSpeed;
+
+
+  if (isPlaying == false && timestampNow !== 0) {
+    admin.firestore().collection('rooms').doc(roomID).collection('messages').add({
+      name: 'Lounge Bot',
+      profilePicUrl: '/images/LoungeLogo.png',
+      text: `Video paused`,
+      timestamp: admin.firestore.FieldValue.serverTimestamp(),
+    });
+  }
+  /** notifies user if video is playing after prevously being paused 
+      except at beginning of video to prevent repeated messages */
+  if (previousVideoState == false && timestampNow > 1) {
+    admin.firestore().collection('rooms').doc(roomID).collection('messages').add({
```

Comment:
except for the text, rest of the functionality is the same in 3 places.  Can we create a util function for it.?